zcash_client_sqlite/src/wallet.rs

```diff
@@ -277,25 +285,73 @@ pub(crate) fn max_zip32_account_index(
     )
 }
 
-/// Returns (account_type, hd_seed_fingerprint, hd_account_index, uvk) for a given account.  
+struct AccountSqlValues<'a> {
+    account_type: u32,
+    hd_seed_fingerprint: Option<&'a [u8]>,
+    hd_account_index: Option<u32>,
+    ufvk: Option<String>,
+    uivk: String,
+}
+
+/// Returns (account_type, hd_seed_fingerprint, hd_account_index, ufvk, uivk) for a given account.  
 fn get_sql_values_for_account_parameters<'a, P: consensus::Parameters>(
     account: &'a Account,
     params: &P,
-) -> (u32, Option<&'a [u8]>, Option<u32>, String) {
-    match account {
-        Account::Zip32(hdaccount) => (
-            AccountType::Zip32.into(),
-            Some(hdaccount.hd_seed_fingerprint().as_bytes()),
-            Some(hdaccount.account_index().into()),
-            hdaccount.ufvk().encode(params),
-        ),
-        Account::Imported(ImportedAccount::Full(ufvk)) => (
-            AccountType::ImportedUfvk.into(),
-            None,
-            None,
-            ufvk.encode(params),
-        ),
+) -> Result<AccountSqlValues<'a>, SqliteClientError> {
+    Ok(match account {
+        Account::Zip32(hdaccount) => AccountSqlValues {
+            account_type: AccountType::Zip32.into(),
+            hd_seed_fingerprint: Some(hdaccount.hd_seed_fingerprint().as_bytes()),
+            hd_account_index: Some(hdaccount.account_index().into()),
+            ufvk: Some(hdaccount.ufvk().encode(params)),
+            uivk: ufvk_to_uivk(hdaccount.ufvk(), params)?,
+        },
+        Account::Imported(ImportedAccount::Full(ufvk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: Some(ufvk.encode(params)),
+            uivk: ufvk_to_uivk(ufvk, params)?,
+        },
+        Account::Imported(ImportedAccount::Incoming(uivk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: None,
+            uivk: uivk.encode(&params.network_type()),
+        },
+    })
+}
+
+pub(crate) fn ufvk_to_uivk<P: consensus::Parameters>(
+    ufvk: &UnifiedFullViewingKey,
+    params: &P,
+) -> Result<String, SqliteClientError> {
+    let mut ivks: Vec<Ivk> = Vec::new();
+    if let Some(orchard) = ufvk.orchard() {
```

Comment:
This is not currently necessary because we always enable `zcash_keys/orchard` for backwards-compatibility. Once we finish Orchard support in `zcash_client_sqlite` it will become necessary, but for now it's fine.